gbus/worker.go

```diff
@@ -265,42 +285,50 @@ func (worker *worker) processMessage(delivery amqp.Delivery, isRPCreply bool) {
 			worker.span.LogFields(slog.String("panic", "failed to process message"))
 			logEntry.Error("failed to process message")
 		}
+		//always finish the span once the stack winds down
 		worker.span.Finish()
 	}()
 
 	worker.log().WithFields(logrus.Fields{"worker": worker.consumerTag, "message_id": delivery.MessageId}).Info("GOT MSG")
 
 	//handle a message that originated from a deadletter exchange
-	if worker.isDead(delivery) {
+	if worker.isDead(delivery) && worker.deadletterHandler != nil {
 		worker.span.LogFields(slog.Error(errors.New("handling dead-letter delivery")))
 		worker.log().Info("invoking deadletter handler")
 		worker.invokeDeadletterHandler(delivery)
 		return
 	}
 
-	bm, err := worker.extractBusMessage(delivery)
-	if err != nil {
-		worker.span.LogFields(slog.Error(err), slog.String("grabbit", "message is poison"))
-		//reject poison message
-		_ = worker.reject(false, delivery)
-		return
+	if err := worker.runGlobalHandler(&delivery); err != nil {
+		//when the global handler fails terminate executation and reject the message
+		_ = worker.reject(true, delivery)
```

Comment:
no return here?

Reply:
No because  worker.reject has a retry strategy that keeps going until it succeeds so there is no error returned...worker.reject has a poor signature and needs to be changed to not return an error